### PR TITLE
Respect unpinned

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 1]
```diff
@@ -412,29 +407,30 @@ data Array a =
...
         { arrContents = contents
         , arrStart = p
         , aEnd   = p
-        , aBound = p `plusPtr` size
+        , aBound = p + count
         }
 
-newAlignedArrayContents :: Int -> Int -> IO (ArrayContents, Ptr a)
+-- The size is in bytes
+newAlignedArrayContents :: Int -> Int -> IO (ArrayContents a, Int)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:422]: `(ArrayContents a, Int)` can be changed to `ArrayContents a`

**Comment by adithyaov** [comment]
```
Streamly/Benchmark/Unicode/Stream.hs:213:1: copyStreamLatin1' `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Unicode/Stream.hs:214:1: copyStreamLatin1' `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState failed:
Streamly/Benchmark/Unicode/Stream.hs:215:1: copyStreamLatin1' `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState failed:
Streamly/Benchmark/Unicode/Stream.hs:217:1: copyStreamLatin1' `hasNoType` Streamly.Internal.Data.Fold.Step.Step failed:
Streamly/Benchmark/Unicode/Stream.hs:218:1: copyStreamLatin1' `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe failed:
Streamly/Benchmark/Unicode/Stream.hs:234:1: copyStreamLatin1 `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Unicode/Stream.hs:235:1: copyStreamLatin1 `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState failed:
Streamly/Benchmark/Unicode/Stream.hs:236:1: copyStreamLatin1 `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState failed:
Streamly/Benchmark/Unicode/Stream.hs:238:1: copyStreamLatin1 `hasNoType` Streamly.Internal.Data.Fold.Type.ManyState failed:
Streamly/Benchmark/Unicode/Stream.hs:239:1: copyStreamLatin1 `hasNoType` Streamly.Internal.Data.Fold.Step.Step failed:
Streamly/Benchmark/Unicode/Stream.hs:240:1: copyStreamLatin1 `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe failed:

Streamly/Benchmark/Prelude/Serial/Exceptions.hs:130:1: readWriteOnExceptionStream does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:141:1: readWriteHandleExceptionStream does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:151:1: readWriteFinally_Stream does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:167:1: fromToBytesBracket_Stream does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:184:1: readWriteBeforeAfterStream `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:193:1: readWriteAfterStream `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:202:1: readWriteAfter_Stream does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:203:1: readWriteAfter_Stream `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:243:1: readChunksOnException does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:253:1: readChunksBracket_ does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Exceptions.hs:287:1: toChunksBracket_ does not contain dictionary values failed:

Streamly/Benchmark/Prelude/Serial/Split.hs:67:1: splitOn does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Split.hs:68:1: splitOn `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:69:1: splitOn `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:70:1: splitOn `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:80:1: splitOnSuffix does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Split.hs:81:1: splitOnSuffix `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:82:1: splitOnSuffix `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:83:1: splitOnSuffix `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:93:1: splitWithSuffix does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Split.hs:94:1: splitWithSuffix `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:95:1: splitWithSuffix `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:96:1: splitWithSuffix `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/Prelude/Serial/Split.hs:124:1: wordsBy does not contain dictionary values failed:
Streamly/Benchmark/Prelude/Serial/Split.hs:125:1: wordsBy `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Prelude/Serial/Split.hs:126:1: wordsBy `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState failed:
Streamly/Benchmark/Prelude/Serial/Split.hs:127:1: wordsBy `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState failed:

Streamly/Benchmark/FileSystem/Handle/Read.hs:239:1: chunksOf `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:238:1: chunksOf `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:237:1: chunksOf `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:236:1: chunksOf `hasNoType` Streamly.Internal.Data.Stream.StreamD.Type.FoldMany passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:235:1: chunksOf `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:234:1: chunksOf does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:197:1: getChunksConcatUnfoldCountLines `hasNoType` Streamly.Internal.Data.Stream.StreamD.Type.ConcatMapUState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:196:1: getChunksConcatUnfoldCountLines `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:195:1: getChunksConcatUnfoldCountLines does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:150:1: readDecodeUtf8 does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:119:1: readSumBytes `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:118:1: readSumBytes `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:117:1: readSumBytes `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:116:1: readSumBytes does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:107:1: readCountWords does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:95:1: readCountLines `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:94:1: readCountLines `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:93:1: readCountLines `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:92:1: readCountLines does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:80:1: readCountBytes `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:79:1: readCountBytes `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:78:1: readCountBytes `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:77:1: readCountBytes does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/Read.hs:68:1: readLast `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:67:1: readLast `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:66:1: readLast `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/Read.hs:65:1: readLast does not contain dictionary values failed:

Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:191:1: writeRead `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:190:1: writeRead `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:189:1: writeRead `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:188:1: writeRead `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:187:1: writeRead does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:174:1: writeReadWith `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:173:1: writeReadWith `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:172:1: writeReadWith `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:171:1: writeReadWith `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:170:1: writeReadWith does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:127:1: readWithFromBytesNull `hasNoType` Streamly.Internal.Data.Stream.StreamD.Type.FoldMany passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:126:1: readWithFromBytesNull `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:125:1: readWithFromBytesNull `hasNoType` Streamly.Internal.Data.Array.Stream.Mut.Foreign.SpliceState passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:124:1: readWithFromBytesNull `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:123:1: readWithFromBytesNull does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:113:1: readFromBytesNull `hasNoType` Streamly.Internal.Data.Stream.StreamD.Type.FoldMany passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:112:1: readFromBytesNull `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:111:1: readFromBytesNull `hasNoType` Streamly.Internal.Data.Array.Stream.Mut.Foreign.SpliceState passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:110:1: readFromBytesNull `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:109:1: readFromBytesNull does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:87:1: copyStream `hasNoType` Streamly.Internal.Data.Tuple.Strict.Tuple3' passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:86:1: copyStream `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ArrayUnsafe failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:85:1: copyStream `hasNoType` Streamly.Internal.Data.Array.Foreign.Mut.Type.ReadUState failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:84:1: copyStream `hasNoType` Streamly.Internal.Data.Unfold.Type.ConcatState failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:83:1: copyStream `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:82:1: copyStream does not contain dictionary values failed:
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:60:1: copyChunks `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/FileSystem/Handle/ReadWrite.hs:59:1: copyChunks does not contain dictionary values failed:

Streamly/Benchmark/Data/Unfold.hs:794:1: readWriteBracket_Unfold does not contain dictionary values failed:
Streamly/Benchmark/Data/Unfold.hs:778:1: readWriteFinally_Unfold does not contain dictionary values failed:
Streamly/Benchmark/Data/Unfold.hs:767:1: readWriteHandleExceptionUnfold does not contain dictionary values failed:
Streamly/Benchmark/Data/Unfold.hs:755:1: readWriteOnExceptionUnfold does not contain dictionary values failed:

Streamly/Benchmark/Data/Array/Stream/Foreign.hs:204:1: copyChunksSplitInterpose `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:203:1: copyChunksSplitInterpose does not contain dictionary values except of Streamly.Internal.Data.Unboxed.Storable failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:189:1: copyChunksSplitInterposeSuffix `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:188:1: copyChunksSplitInterposeSuffix does not contain dictionary values except of Streamly.Internal.Data.Unboxed.Storable failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:149:1: toChunksSplitOn `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:148:1: toChunksSplitOn does not contain dictionary values failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:139:1: toChunksSplitOnSuffix `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:138:1: toChunksSplitOnSuffix does not contain dictionary values failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:124:1: toChunksDecodeUtf8Arrays does not contain dictionary values failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:116:1: toChunksCountBytes `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:115:1: toChunksCountBytes does not contain dictionary values failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:104:1: toChunksSumLengths `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:103:1: toChunksSumLengths does not contain dictionary values failed:
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:93:1: toChunksLast `hasNoType` Streamly.Internal.Data.Stream.StreamD.Step.Step passed.
Streamly/Benchmark/Data/Array/Stream/Foreign.hs:92:1: toChunksLast does not contain dictionary values failed:
```

**Comment by adithyaov** [comment]
Benchmarks that need to be checked:
```
decode-encode/encodeLatin1' . decodeLatin1
decode-encode/encodeLatin1 . decodeLatin1

exceptions/S.onException
exceptions/S.handle
exceptions/S.finally_
exceptions/S.bracket_
exceptions/S.after . S.before
exceptions/S.after
exceptions/S.after_

exceptions/readChunks/UF.onException
exceptions/readChunks/UF.bracket_

exceptions/toChunks/S.bracket_

split/S.splitOn (== lf) FL.drain
split/S.splitWithSuffix (== lf) FL.drain
split/S.splitWithSuffixSeq \"\\r\\n\" FL.drain
split/S.wordsBy isSpace FL.drain

reduce/read/chunks/S.chunksOf 1
reduce/read/chunks/S.chunksOf 10
reduce/read/chunks/S.chunksOf 1000

reduce/toBytes/US.lines . SS.decodeLatin1

reduce/read/SS.decodeUtf8
reduce/read/S.length
reduce/read/US.lines . SS.decodeLatin1
reduce/read/US.words . SS.decodeLatin1
reduce/read/S.last
reduce/read/S.sum

copy/FH.write . FH.read
copy/FH.writeWith . FH.readWith

copy/putBytes/FH.readWith
copy/putBytes/rawToNull

copy/read/rawToFile
copy/read/rawToNull

copy/getChunks/toNull
copy/getChunks/raw

exceptions/UF.onException
exceptions/UF.handle
exceptions/UF.finally_
exceptions/UF.bracket_

reduce/toChunks/Stream.last
reduce/toChunks/Stream.sum . Stream.map Array.length
reduce/toChunks/countBytes
reduce/toChunks/decodeUtf8Arrays
reduce/toChunks/splitOnSuffix
reduce/toChunks/splitOn

copy/toChunks/group-ungroup/interposeSuffix . splitOnSuffix
copy/toChunks/group-ungroup/interpose . splitOn
```

File: core/src/Streamly/Internal/Data/Ring/Foreign.hs [hunk 2]
```diff
@@ -426,12 +426,12 @@ unsafeEqArrayN Ring{..} rh A.Array{..} n =
 -- supplied array must be equal to or bigger than the ringBuffer, ARRAY BOUNDS
 -- ARE NOT CHECKED.
 {-# INLINE unsafeEqArray #-}
-unsafeEqArray :: Ring a -> Ptr a -> A.Array a -> Bool
-unsafeEqArray Ring{..} rh A.Array{..} =
-    let !res = unsafeInlineIO $ do
+unsafeEqArray :: forall a. Storable a => Ring a -> Ptr a -> A.Array a -> Bool
+unsafeEqArray Ring{..} rh arr@(A.Array{..}) =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Ring/Foreign.hs:430]: We should be able to compare it with an unpinned array as well.

File: core/src/Streamly/Internal/Data/Ring/Foreign.hs [hunk 3]
```diff
@@ -400,12 +400,12 @@ cast arr =
 -- the ring buffer. This is unsafe because the ringHead Ptr is not checked to
 -- be in range.
 {-# INLINE unsafeEqArrayN #-}
-unsafeEqArrayN :: Ring a -> Ptr a -> A.Array a -> Int -> Bool
-unsafeEqArrayN Ring{..} rh A.Array{..} n =
-    let !res = unsafeInlineIO $ do
+unsafeEqArrayN :: forall a. Storable a => Ring a -> Ptr a -> A.Array a -> Int -> Bool
+unsafeEqArrayN Ring{..} rh arr@(A.Array{..}) n =
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Ring/Foreign.hs:404]: We should be able to compare it with an unpinned array as well.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 4]
```diff
@@ -793,36 +805,45 @@ arrayChunkBytes = 1024
...
 roundDownTo elemSize size = size - (size `mod` elemSize)
 
+-- The size is in bytes
 -- XXX See if resizing can be implemented by reading the old array as a stream
 -- and then using writeN to the new array.
 --
 {-# NOINLINE reallocAligned #-}
 reallocAligned :: Int -> Int -> Int -> Array a -> IO (Array a)
-reallocAligned elemSize alignSize newCapacity Array{..} = do
+reallocAligned elemSize alignSize newCapacityInBytes Array{..} = do
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:814]: Please look at this a little more carefully.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 5]
```diff
@@ -846,22 +868,27 @@ reallocWith :: forall m a. (MonadIO m , Storable a) =>
     -> Array a
     -> m (Array a)
 reallocWith label capSizer minIncr arr = do
-    let oldSize = aEnd arr `minusPtr` arrStart arr
-        newCap = capSizer oldSize
-        newSize = oldSize + minIncr
-        safeCap = max newCap newSize
-    assert (newCap >= newSize || error (badSize newSize)) (return ())
-    realloc safeCap arr
+    let oldSize = aEnd arr - arrStart arr
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:871]: Little more careful review.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 6]
```diff
@@ -911,14 +939,15 @@ rightSize :: forall m a. (MonadIO m, Storable a) => Array a -> m (Array a)
 rightSize arr@Array{..} = do
     assert (aEnd <= aBound) (return ())
     let start = arrStart
-        len = aEnd `minusPtr` start
-        capacity = aBound `minusPtr` start
-        target = roundUpLargeArray len
-        waste = aBound `minusPtr` aEnd
-    assert (target >= len) (return ())
-    assert (len `mod` SIZE_OF(a) == 0) (return ())
+        len = aEnd - start
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:942]: Little more careful review.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 7]
```diff
@@ -2157,39 +2188,45 @@ fromListRev xs = fromListRevN (Prelude.length xs) xs
...
 {-# INLINE spliceCopy #-}
-spliceCopy :: (MonadIO m, Storable a) => Array a -> Array a -> m (Array a)
-spliceCopy arr1 arr2 = do
-    let src1 = arrStart arr1
-        src2 = arrStart arr2
-        len1 = aEnd arr1 `minusPtr` src1
-        len2 = aEnd arr2 `minusPtr` src2
+spliceCopy :: forall m a. (MonadIO m, Storable a) => Array a -> Array a -> m (Array a)
+spliceCopy arr1 arr2 = liftIO $ do
+    let start1 = arrStart arr1
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:2193]: Careful review.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 8]
```diff
@@ -2157,39 +2188,45 @@ fromListRev xs = fromListRevN (Prelude.length xs) xs
...
+spliceUnsafe dst (src, srcLenInBytes) =
     liftIO $ do
-         let psrc = arrStart src
-         let pdst = aEnd dst
-         assert (pdst `plusPtr` srcLen <= aBound dst) (return ())
-         memcpy (castPtr pdst) (castPtr psrc) srcLen
-         touch (arrContents src)
-         touch (arrContents dst)
-         return $ dst {aEnd = pdst `plusPtr` srcLen}
+         let startSrc = arrStart src
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:2218]: Careful review.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 9]
```diff
@@ -2320,72 +2358,108 @@ splitAt i arr@Array{..} =
...
-#ifdef DEVBUILD
-    Storable b =>
-#endif
-    Array a -> Array b
+       forall a b. (Storable a, Storable b)
+    => Array a
+    -> Array b
 castUnsafe (Array contents start end bound) =
-    Array contents (castPtr start) (castPtr end) (castPtr bound)
+    let sizeB = SIZE_OF(b)
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:2365]: Careful review

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 10]
```diff
@@ -2320,72 +2358,108 @@ splitAt i arr@Array{..} =
...
-        then
-            if ptr1 == ptr2
-            then return True
-            else do
-                r <- memcmp (castPtr ptr1) (castPtr ptr2) len1
-                touch (arrContents arr1)
-                touch (arrContents arr2)
-                return r
-        else return False
+    liftIO
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:2442]: Careful review

File: core/src/Streamly/Internal/Data/Ring/Foreign.hs (hunk 3 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Ring/Foreign.hs:404]: We should always peek and compare?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
I have added a commit with review comments.

File: core/src/Streamly/Internal/Data/Unboxed.hs [hunk 11]
```diff
@@ -74,11 +75,22 @@ touch (ArrayContents contents) =
...
 class Unboxed a where
     sizeOf :: a -> Int
     alignment :: a -> Int
+
+    -- Read an element of type "a" from a MutableByteArray given the element
+    -- index.
     readByteArray :: ArrayContents a -> Int -> IO a
     writeByteArray :: ArrayContents a -> Int -> a -> IO ()
+
+    -- Read an element of type "a" from a Ptr given the element index.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Unboxed.hs:93]: Now that we use elements, it's pretty weird to use `Ptr` to consider element index.

File: src/Streamly/Internal/Unicode/Stream.hs [hunk 12]
```diff
@@ -237,18 +237,21 @@ decodeTable = [
...
-    -- Aligning to cacheline makes a barely noticeable difference
-    -- XXX currently alignment is not implemented for unmanaged allocation
     let size = utf8dLength
     p <- liftIO $ mallocBytes size
     void $ D.fold
         (Fold.foldlM' (\b a -> poke b a >> return (b `plusPtr` 1)) (return p))
         (D.fromList decodeTable)
     return p
 
+-- XXX We can just use Foreign.Storable in this.
```
Comment (adithyaov) [inline src/Streamly/Internal/Unicode/Stream.hs:254]: We could, but there was a specific reason for changing this, which I seem to have forgotten. Either way, better to use one class for unboxed representation.

**Pull request closed without merging**